### PR TITLE
Add withStringBody method

### PR DESCRIPTION
**Pull request by cleptric**
This implements #10290 

Is sending an empty response body if no string was passed a good idea?

### CODE REVIEW

File: src/Http/Response.php [hunk 1]
```diff
@@ -2233,6 +2233,30 @@ public function withFile($path, array $options = [])
...
+
+        if (is_string($string) === false) {
+            if ($body->isSeekable()) {
+                $body->rewind();
+            }
+            $body->write('');
+
+            return $new;
+        }
+        $body->write($string);
```
Comment (markstory) [inline src/Http/Response.php]: You could replace the entire stream with a new instance as this method is explicitly replacing the body.

File: tests/TestCase/Network/ResponseTest.php [hunk 2]
```diff
@@ -2778,6 +2778,49 @@ public function testWithBody()
...
+     * @return void
+     */
+    public function testWithStringBody()
+    {
+        $response = new Response();
+        $response->withStringBody('Foo');
+        $body = $response->getBody();
+        $body->rewind();
+        $result = $body->getContents();
+        $this->assertSame('Foo', $result);
```
Comment (markstory) [inline tests/TestCase/Network/ResponseTest.php]: String casting a stream should rewind and get the contents. That should help make the tests simpler.

File: tests/TestCase/Network/ResponseTest.php [hunk 3]
```diff
@@ -2778,6 +2778,49 @@ public function testWithBody()
...
+
+        $response = new Response();
+        $response->withStringBody(1337);
+        $body = $response->getBody();
+        $body->rewind();
+        $result = $body->getContents();
+        $this->assertSame('', $result);
+
+        $response = new Response();
+        $response->withStringBody(['foo' => 'bar']);
```
Comment (markstory) [inline tests/TestCase/Network/ResponseTest.php]: Should this be an error instead?

File: tests/TestCase/Network/ResponseTest.php [hunk 4]
```diff
@@ -2778,6 +2778,49 @@ public function testWithBody()
...
+
+        $response = new Response();
+        $response->withStringBody('');
+        $body = $response->getBody();
+        $body->rewind();
+        $result = $body->getContents();
+        $this->assertSame('', $result);
+
+        $response = new Response();
+        $response->withStringBody(null);
```
Comment (markstory) [inline tests/TestCase/Network/ResponseTest.php]: What about objects that have a __toString() method?

File: src/Http/Response.php [hunk 5]
```diff
@@ -2233,6 +2233,30 @@ public function withFile($path, array $options = [])
...
+     *
+     * @param string $string The string to be sent
+     * @return static
+     */
+    public function withStringBody($string)
+    {
+        $new = clone $this;
+        $body = $new->getBody();
+
+        if (is_string($string) === false) {
```
Comment (ADmad) [inline src/Http/Response.php]: How about we just typecast the `$string` argument to string?

File: tests/TestCase/Network/ResponseTest.php (hunk 4 above)
Comment (ADmad) [inline tests/TestCase/Network/ResponseTest.php]: As I suggested above we should just typecast the argument to string.

**Comment by cleptric** [comment]
@ADmad @markstory Thank you for your good advice 👍

File: tests/TestCase/Network/ResponseTest.php [hunk 6]
```diff
@@ -2778,6 +2778,45 @@ public function testWithBody()
...
+        $this->assertSame('1337', (string)$body);
+
+        /*
+         * This should produce an "Array to string conversion" error
+         * which gets thrown as a \PHPUnit\Framework\Error\Error Exception by PHPUnit.
+         */
+        $response = new Response();
+        $response->withStringBody(['foo' => 'bar']);
+        $body = $response->getBody();
+        $this->assertSame('', (string)$body);
```
Comment (markstory) [inline tests/TestCase/Network/ResponseTest.php]: You could make this a separate case. Right now the expected exception could mask failure elsewhere in the test.

Comment (cleptric) [inline tests/TestCase/Network/ResponseTest.php]: Good point' I'll change this.

File: src/Http/Response.php [hunk 7]
```diff
@@ -2233,6 +2233,21 @@ public function withFile($path, array $options = [])
     }
 
     /**
+     * Convenience method to set a string into the response body
+     *
+     * @param string $string The string to be sent
+     * @return static
```
Comment (ADmad) [inline src/Http/Response.php]: Should be `@return $this`

File: src/Http/Response.php [hunk 8]
```diff
@@ -2233,6 +2233,21 @@ public function withFile($path, array $options = [])
...
+     */
+    public function withStringBody($string)
+    {
+        $new = clone $this;
+        $new->_createStream();
+        $this->stream->write((string)$string);
```
Comment (saeideng) [inline src/Http/Response.php]: `$this` or `$new` ?

File: tests/TestCase/Network/ResponseTest.php [hunk 9]
```diff
@@ -2778,6 +2778,52 @@ public function testWithBody()
...
+        $response = new Response();
+        $response->withStringBody(null);
+        $body = $response->getBody();
+        $this->assertSame('', (string)$body);
+
+        $response = new Response();
+        $response->withStringBody(1337);
+        $body = $response->getBody();
+        $this->assertSame('1337', (string)$body);
+    }
```
Comment (saeideng) [inline tests/TestCase/Network/ResponseTest.php]: also you need to ensure you created new obj with `withStringBody()` but i not see this assert

File: src/Http/Response.php [hunk 10]
```diff
@@ -2233,6 +2233,21 @@ public function withFile($path, array $options = [])
...
     /**
+     * Convenience method to set a string into the response body
+     *
+     * @param string $string The string to be sent
+     * @return static
+     */
+    public function withStringBody($string)
+    {
+        $new = clone $this;
+        $new->_createStream();
```
Comment (ADmad) [inline src/Http/Response.php]: `_createStream()` is a protected method. PHP still allows it to be used just because you are in instance of same class?

Comment (josegonzalez) [inline src/Http/Response.php]: Yeah, why wouldn't that work? It's a subclass calling a `private` method that won't work.

Comment (ADmad) [inline src/Http/Response.php]: A separate instance `$new` is being used not `$this`.

Comment (ADmad) [inline src/Http/Response.php]: My understanding of OOP encapsulation is that no one except the object/instance itself should be able to access it's protected methods.

Comment (josegonzalez) [inline src/Http/Response.php]: Ah I didn't see that. Nvm me :P

Comment (saeideng) [inline src/Http/Response.php]: $this and $new used some class and this method used within this class

Comment (ADmad) [inline src/Http/Response.php]: > $this and $new used some class and this method used within this class

Yes I know that but I find it strange PHP allows accessing protected method of other objects just cause you are trying to access it from same class. Wonder how other languages handle such situation.

Comment (lorenzo) [inline src/Http/Response.php]: It is not strange, that is the design of that class visibility. As a fun fact, protected in java means that any class in the same package (namespace) can access that property or method

Comment (ADmad) [inline src/Http/Response.php]: http://i.giphy.com/l0MYEqEzwMWFCg8rm.gif

File: src/Http/Response.php (hunk 8 above)
Comment (markstory) [inline src/Http/Response.php]: This should be `$new` not `$this`.

File: tests/TestCase/Network/ResponseTest.php [hunk 11]
```diff
@@ -2778,6 +2778,52 @@ public function testWithBody()
...
+    public function testWithStringBody()
+    {
+        $response = new Response();
+        $response->withStringBody('Foo');
+        $body = $response->getBody();
+        $this->assertSame('Foo', (string)$body);
+
+        $response = new Response();
+        $response->withStringBody('');
+        $body = $response->getBody();
```
Comment (markstory) [inline tests/TestCase/Network/ResponseTest.php]: These tests are checking the wrong response object. They should be checking the return value of `withStringBody()` like:

```php
$response = new Response();
$response = $response->withStringBody('');
$body = $response->getBody();
$this->assertSame('', (string)$body);
```

Right now your tests are checking the original response which is not how the immutable style methods should work.

File: src/Http/Response.php (hunk 7 above)
Comment (cleptric) [inline src/Http/Response.php]: The other methods are `@return static`

File: src/Http/Response.php (hunk 8 above)
Comment (cleptric) [inline src/Http/Response.php]: $new!

File: src/Http/Response.php (hunk 7 above)
Comment (ADmad) [inline src/Http/Response.php]: My bad. `@return static` is correct due to immutability.

**Comment by markstory** [comment]
Nice work @cleptric

**Comment by cleptric** [comment]
Docs/migration guide will follow

**Pull request merged**